File: prompts/template-loader.js
#!/usr/bin/env node
/**
 * Prompt Template Loader for Node.js
 * تحميل واستخدام قوالب Prompts بسهولة
 */

import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * محمّل قوالب Prompts مع دعم الاستبدال التلقائي
 */
export class PromptTemplateLoader {
  constructor(templatesPath = 'prompts/prompt-templates.json') {
    this.templatesPath = templatesPath;
    this.templates = null;
    this.agent = null;
  }

  /**
   * تحميل القوالب من الملف
   */
  async loadTemplates() {
    const data = await fs.readFile(this.templatesPath, 'utf-8');
    const parsed = JSON.parse(data);
    this.templates = {};

    for (const template of parsed.templates) {
      this.templates[template.id] = template;
    }

    return this.templates;
  }

  /**
   * الحصول على قالب بواسطة ID
   */
  getTemplate(templateId) {
    if (!this.templates) {
      throw new Error('Templates not loaded. Call loadTemplates() first.');
    }
    return this.templates[templateId] || null;
  }

  /**
   * عرض جميع القوالب أو القوالب الخاصة بـ integration معين
   */
  listTemplates(integration = null) {
    if (!this.templates) {
      throw new Error('Templates not loaded. Call loadTemplates() first.');
    }

    const allTemplates = Object.values(this.templates);

    if (integration) {
      return allTemplates.filter(t => t.integration === integration);
    }

    return allTemplates;
  }

  /**
   * تطبيق القالب مع استبدال المتغيرات
   */
  renderPrompt(templateId, variables = {}) {
    const template = this.getTemplate(templateId);

    if (!template) {
      throw new Error(`Template '${templateId}' not found`);
    }

    let systemMessage = template.system_message;
    let userMessage = template.user_message;

    // استبدال placeholders
    for (const [key, value] of Object.entries(variables)) {
      const placeholder = `{{${key}}}`;
      userMessage = userMessage.replaceAll(placeholder, String(value));
    }

    return {
      system: systemMessage,
      user: userMessage,
      template_id: templateId,
      integration: template.integration
    };
  }

  /**
   * تنفيذ قالب مباشرة مع Agent
   */
  async executeTemplate(templateId, agent, variables = {}) {
    const template = this.getTemplate(templateId);

    if (!template) {
      throw new Error(`Template '${templateId}' not found`);
    }

    // إضافة الـ MCP server المطلوب
    const integration = template.integration;
    const serverUrl = this.getSkillUrl(integration);

    await agent.addMCPServer({
      name: integration,
      url: serverUrl,
      type: 'url'
    });

    // تطبيق القالب
    const rendered = this.renderPrompt(templateId, variables);

    // تنفيذ
    const response = await agent.processMessage(rendered.user);

    return response;
  }

  /**
   * الحصول على URL الخاص بـ integration
   */
  getSkillUrl(integration) {
    const urls = {
      'google-workspace': 'http://localhost:3001',
      'notion': 'http://localhost:3002',
      'airtable': 'http://localhost:3003',
      'telegram': 'http://localhost:3004',
      'whatsapp': 'http://localhost:3005',
      'twitter': 'http://localhost:3006',
      'reddit': 'http://localhost:3007',
      'filesystem': 'http://localhost:3008'
    };

    return urls[integration] || 'http://localhost:3001';
  }

  /**
   * الحصول على تقدير التوكنات للقالب
   */
  getTokenEstimate(templateId) {
    const template = this.getTemplate(templateId);
    return template ? template.token_estimate || 0 : 0;
  }

  /**
   * الحصول على نصائح توفير التوكنات
   */
  getTips(templateId) {
    const template = this.getTemplate(templateId);
    return template ? template.token_saving_tips || [] : [];
  }

  /**
   * عرض معلومات القالب
   */
  showTemplateInfo(templateId) {
    const template = this.getTemplate(templateId);

    if (!template) {
      console.log(`❌ Template '${templateId}' not found`);
      return;
    }

    console.log(`\n📋 Template: ${template.id}`);
    console.log(`🔷 Integration: ${template.integration}`);
    console.log(`📝 Intent: ${template.intent}`);
    console.log(`💡 Token Estimate: ~${template.token_estimate} tokens`);
    console.log(`\n🎯 Token Saving Tips:`);
    template.token_saving_tips.forEach(tip => {
      console.log(`  • ${tip}`);
    });

    if (template.notes) {
      console.log(`\n⚠️  Notes: ${template.notes}`);
    }
  }
}

// أمثلة للاستخدام
async function examples() {
  const loader = new PromptTemplateLoader();
  await loader.loadTemplates();

  // مثال 1: عرض جميع قوالب Google Workspace
  console.log('📧 Google Workspace Templates:');
  const gwTemplates = loader.listTemplates('google-workspace');
  gwTemplates.forEach(t => {
    console.log(`  - ${t.id}: ${t.intent}`);
  });

  console.log('\n' + '='.repeat(50) + '\n');

  // مثال 2: تطبيق قالب
  const rendered = loader.renderPrompt('gw_create_calendar_event', {
    title: 'اجتماع الفريق',
    date: '2025-12-01',
    time: '14:00',
    duration: '60min',
    attendees: 'team@company.com',
    description: 'مراجعة Q1'
  });

  console.log('📋 Rendered Prompt:');
  console.log(`System: ${rendered.system}`);
  console.log(`User: ${rendered.user}`);

  console.log('\n' + '='.repeat(50) + '\n');

  // مثال 3: معلومات القالب
  loader.showTemplateInfo('notion_create_page');

  console.log('\n' + '='.repeat(50) + '\n');

  // مثال 4: قائمة جميع الـ integrations
  const integrations = [...new Set(
    loader.listTemplates().map(t => t.integration)
  )];

  console.log('🔌 Available Integrations:');
  integrations.forEach(i => {
    const count = loader.listTemplates(i).length;
    console.log(`  • ${i}: ${count} templates`);
  });
}

// تشغيل الأمثلة إذا تم تنفيذ الملف مباشرة
if (import.meta.url === `file://${process.argv[1]}`) {
  examples().catch(console.error);
}

export default PromptTemplateLoader;

File: prompts/template_loader.py
#!/usr/bin/env python3
"""
Prompt Template Loader for Python
تحميل واستخدام قوالب Prompts بسهولة
"""

import json
import os
from typing import Dict, List, Optional, Any
from anthropic_skills import AnthropicSkills


class PromptTemplateLoader:
    """محمّل قوالب Prompts مع دعم الاستبدال التلقائي"""

    def __init__(self, templates_path: str = "prompts/prompt-templates.json"):
        """
        تهيئة المحمّل

        Args:
            templates_path: مسار ملف القوالب
        """
        self.templates_path = templates_path
        self.templates = self._load_templates()
        self.claude = None

    def _load_templates(self) -> Dict[str, Any]:
        """تحميل القوالب من الملف"""
        with open(self.templates_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
            return {t['id']: t for t in data['templates']}

    def get_template(self, template_id: str) -> Optional[Dict]:
        """
        الحصول على قالب بواسطة ID

        Args:
            template_id: معرّف القالب

        Returns:
            القالب أو None
        """
        return self.templates.get(template_id)

    def list_templates(self, integration: Optional[str] = None) -> List[Dict]:
        """
        عرض جميع القوالب أو القوالب الخاصة بـ integration معين

        Args:
            integration: اسم الـ integration (اختياري)

        Returns:
            قائمة القوالب
        """
        if integration:
            return [t for t in self.templates.values() if t['integration'] == integration]
        return list(self.templates.values())

    def render_prompt(self, template_id: str, **kwargs) -> Dict[str, str]:
        """
        تطبيق القالب مع استبدال المتغيرات

        Args:
            template_id: معرّف القالب
            **kwargs: المتغيرات للاستبدال

        Returns:
            dict مع system_message و user_message جاهزين
        """
        template = self.get_template(template_id)
        if not template:
            raise ValueError(f"Template '{template_id}' not found")

        system_message = template['system_message']
        user_message = template['user_message']

        # استبدال placeholders
        for key, value in kwargs.items():
            placeholder = f"{{{{{key}}}}}"
            user_message = user_message.replace(placeholder, str(value))

        return {
            'system': system_message,
            'user': user_message,
            'template_id': template_id,
            'integration': template['integration']
        }

    def execute_template(
        self,
        template_id: str,
        api_key: Optional[str] = None,
        model: str = "claude-sonnet-4-5",
        max_tokens: int = 4096,
        **kwargs
    ) -> str:
        """
        تنفيذ قالب مباشرة مع Claude

        Args:
            template_id: معرّف القالب
            api_key: مفتاح Anthropic API (اختياري)
            model: النموذج المستخدم
            max_tokens: حد التوكنات
            **kwargs: متغيرات القالب

        Returns:
            رد Claude
        """
        # تهيئة Claude إذا لم يكن مهيأً
        if not self.claude:
            self.claude = AnthropicSkills(api_key=api_key)

        template = self.get_template(template_id)
        if not template:
            raise ValueError(f"Template '{template_id}' not found")

        # إضافة الـ skill المطلوب
        integration = template['integration']
        skill_url = self._get_skill_url(integration)

        self.claude.add_skill({
            'name': integration,
            'type': 'url',
            'url': skill_url
        })

        # تطبيق القالب
        rendered = self.render_prompt(template_id, **kwargs)

        # تنفيذ
        response = self.claude.chat(
            prompt=rendered['user'],
            system=rendered['system'],
            model=model,
            max_tokens=max_tokens
        )

        return response

    def _get_skill_url(self, integration: str) -> str:
        """الحصول على URL الخاص بـ integration"""
        urls = {
            'google-workspace': 'http://localhost:3001',
            'notion': 'http://localhost:3002',
            'airtable': 'http://localhost:3003',
            'telegram': 'http://localhost:3004',
            'whatsapp': 'http://localhost:3005',
            'twitter': 'http://localhost:3006',
            'reddit': 'http://localhost:3007',
            'filesystem': 'http://localhost:3008'
        }
        return urls.get(integration, 'http://localhost:3001')

    def get_token_estimate(self, template_id: str) -> int:
        """الحصول على تقدير التوكنات للقالب"""
        template = self.get_template(template_id)
        return template.get('token_estimate', 0) if template else 0

    def get_tips(self, template_id: str) -> List[str]:
        """الحصول على نصائح توفير التوكنات"""
        template = self.get_template(template_id)
        return template.get('token_saving_tips', []) if template else []


# أمثلة للاستخدام
if __name__ == '__main__':
    # تهيئة المحمّل
    loader = PromptTemplateLoader()

    # مثال 1: عرض جميع قوالب Google Workspace
    print("📧 Google Workspace Templates:")
    for template in loader.list_templates('google-workspace'):
        print(f"  - {template['id']}: {template['intent']}")

    print("\n" + "="*50 + "\n")

    # مثال 2: تطبيق قالب
    rendered = loader.render_prompt(
        'gw_create_calendar_event',
        title='اجتماع الفريق',
        date='2025-12-01',
        time='14:00',
        duration='60min',
        attendees='team@company.com',
        description='مراجعة Q1'
    )

    print("📋 Rendered Prompt:")
    print(f"System: {rendered['system']}")
    print(f"User: {rendered['user']}")

    print("\n" + "="*50 + "\n")

    # مثال 3: تقدير التوكنات
    estimate = loader.get_token_estimate('gw_create_calendar_event')
    print(f"💡 Token Estimate: ~{estimate} tokens")

    # مثال 4: نصائح توفير التوكنات
    tips = loader.get_tips('gw_create_calendar_event')
    print("\n🎯 Token Saving Tips:")
    for tip in tips:
        print(f"  • {tip}")

    print("\n" + "="*50 + "\n")

    # مثال 5: تنفيذ مباشر (يحتاج API key)
    # response = loader.execute_template(
    #     'gw_send_email_draft',
    #     to='client@example.com',
    #     subject='متابعة',
    #     body='شكراً للاجتماع',
    #     cc=''
    # )
    # print(f"Response: {response}")
